File: ValidParanth.java
//Time Complexity: O(n) 
//Space Complexity: O(n) 

class Solution {
    public boolean isValid(String s) {
        if(s.length() == 0 || s == null)
            return true;

        if(s.length()%2!=0) return false;
        Stack<Character> st = new Stack<>();

        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);

            if(c == '(') st.push(')');
            else if(c == '{') st.push('}');
            else  if(c == '[') st.push(']');
            else if(st.isEmpty() || c != st.pop()) return false;

        }

        return st.isEmpty();
    }
} 

File: validParanth.py
#TC: O(n)
#SC: O(n)

class Solution:
    def isValid(self, s: str) -> bool:
        p = {'}':'{', ']':'[', ')':'('}
        stack=[]

        for i in s:
            if i in p.values():     
                stack.append(i)
            if i in p.keys():           
                if (stack==[] or stack[-1] != p[i]):
                    return False
                else:
                    stack.pop()

        return stack==[] 